### DISCUSSION TITLE
What is the difference between Workqueue Threads and Message Queues

### DISCUSSION THREAD

**Discussion by mfinmuch** [zephyrproject-rtos/zephyr#30234] [Q&A] [answered]
Hi,
Recently I want to share data in two threads
Use queue communication between two threads
A thread waits for the event of the queue (for example, there is already data in the queue, and it will read the queue when notified)
Another thread generates data and puts it into the queue.
Is there any recommended way?
In addition, when looking for ways, I found **Workqueue Threads** and **Message Queues**. Is there any difference between these two methods? Personally, I think these two methods are quite similar.

thanks,
PYL

**Comment by henrikbrixandersen** [comment, marked as answer]
There are multiple ways of passing data between two threads in Zephyr. The documentation has a nice overview of the various data passing kernel services along with their high level features: https://docs.zephyrproject.org/latest/reference/kernel/index.html#data-passing

Workqueue threads (https://docs.zephyrproject.org/latest/reference/kernel/threads/workqueue.html) and Message Queues (https://docs.zephyrproject.org/latest/reference/kernel/data_passing/message_queues.html) are two complete different things, although their names may sound similar.
